src/block_data.rs

```diff
@@ -0,0 +1,320 @@
+                .try_into()
+                .map_err(|_| BlockHeaderConversionError::BaseFeePerGas)?,
+        );
+        let gas_used = u64::from_le_bytes(
+            DatumIndexV2::gas_used(data_slice)
+                .try_into()
+                .map_err(|_| BlockHeaderConversionError::GasUsed)?,
+        );
+        let logs_bloom = DatumIndexV2::logs_bloom(data_slice)
+            .try_into()
+            .map_err(|_| BlockHeaderConversionError::LogsBloom)?;
+
+
+        Ok(BlockHeaderDataV2 {
+            chain_id,
+            proposer,
+            timestamp,
+            transactions_hash,
+            state_hash,
+            receipts_hash,
+            base_fee_per_gas,
+            gas_used,
+            logs_bloom,
+        })
+    }
+}
+
+/// Enumerates errors in conversion from HotStuffData to BlockDataHeader.
+#[derive(Debug)]
+pub enum BlockHeaderConversionError {
+    /// Wrong number of slice of bytes
+    NumberOfSlots,
+    /// Fail to convert bytes into Chain ID
+    ChainID,
+    /// Fail to convert bytes into Proposer
+    Proposer,
+    /// Fail to convert bytes into Timestamp
+    Timestamp,
+    /// Fail to convert bytes into Transactions Hash
+    TxsHash,
+    /// Fail to convert bytes into State Hash
+    StateHash,
+    /// Fail to convert bytes into Receipt Hash
+    ReceiptsHash,
+    /// Fail to convert bytes into Base Fee Per Gas
+    BaseFeePerGas,
+    /// Fail to convert bytes into Gas Used
+    GasUsed,
+    /// Fail to convert bytes into Logs Bloom
+    LogsBloom,
+}
+
+impl From<BlockHeaderDataV2> for Data {
+    fn from(value: BlockHeaderDataV2) -> Data {
+        let mut buf = vec![Vec::new(); DatumIndexV2::BlockHeaderSize as usize];
+        DatumIndexV2::set_chain_id(&mut buf, value.chain_id.to_le_bytes().to_vec());
+        DatumIndexV2::set_proposer(&mut buf, value.proposer.to_vec());
+        DatumIndexV2::set_timestamp(&mut buf, value.timestamp.to_le_bytes().to_vec());
+        DatumIndexV2::set_transactions_hash(&mut buf, value.transactions_hash.to_vec());
+        DatumIndexV2::set_state_hash(&mut buf, value.state_hash.to_vec());
+        DatumIndexV2::set_receipts_hash(&mut buf, value.receipts_hash.to_vec());
+        DatumIndexV2::set_base_fee_per_gas(&mut buf, value.base_fee_per_gas.to_le_bytes().to_vec());
+        DatumIndexV2::set_gas_used(&mut buf, value.gas_used.to_le_bytes().to_vec());
+        DatumIndexV2::set_logs_bloom(&mut buf, value.logs_bloom.to_vec());
+        buf
+    }
+
+}
+
+
+/// Enumerates errors in conversion from HotStuffData to BlockData.
+#[derive(Debug)]
+pub enum BlockConversionError {
+    /// Fail to convert Block Header.
+    WrongHeader(BlockHeaderConversionError),
+    /// Wrong number of slice of bytes. It should contain equal number of transactions
+    /// and receipts.
+    IncorrectNumberOfTxsAndReceipts,
+    /// Fail to deserialize a transaction.
+    Transaction,
+    /// Signature of a transaction is invalid.
+    InvalidTransactionSignature,
```

Comment:
Does this not correspond to `CryptographicallyIncorrectTransactionError`, which also captures the scenario that a transaction’s hash (not its signature) is wrong?

Reply:
Yes. If the hash of computed signature is not as same as the hash specified in the transaction, this error is returned.